### PR TITLE
feat: configurable s2i builder images

### PR DESCRIPTION
**Pull request by lkingland**
# Changes

- :gift: configurable s2i builder images
- :gift: refactor to enable fuller testing
- :gift: tests

S2I builder can now have its builder image defined in func.yaml (on the Function structure).   Previously it was a shared setting with Pack builders, which meant defining a builder image other than the default would peg the function to only work with one builder or the other.  also enables per-builder settings, which contributes to (but does not fully comple) #1022 

/kind enhancement

### CODE REVIEW

**Review by zroubalik — APPROVED** [review]
/lgtm
/approve

nice!

**Comment by lance** [comment]
/override cla/google since this PR is already covered by easycla

File: s2i/builder.go [hunk 1]
```diff
@@ -120,15 +134,41 @@ func (b *Builder) Build(ctx context.Context, f fn.Function) (err error) {
...
+// Function's Builders map.
+func builderImage(f fn.Function) (string, error) {
 	if f.Runtime == "" {
 		return "", ErrRuntimeRequired
 	}
-	v, ok := DefaultBuilderImages[f.Runtime]
-	if !ok {
-		return "", ErrRuntimeNotSupported
+
+	v, ok := f.BuilderImages["s2i"]
```
Comment (lance) [inline s2i/builder.go:148]: @lkingland am I reading this correctly that if there is an s2i builder image defined, it will be used as the default?

**Pull request merged**